src/components/accordion/accordion_context.provider.tsx

```diff
@@ -1,53 +1,125 @@
 import type { ReactNode } from 'react';
 import { useMemo, useReducer } from 'react';
+import { match } from 'ts-pattern';
 
 import type { ContextType } from './accordion_context.js';
 import { accordionContext } from './accordion_context.js';
-import { reducer } from './accordion_context.state.js';
+import type {
+  AccordionItemState,
+  ActionType,
+} from './accordion_context.utils.js';
+import { getItem } from './accordion_context.utils.js';
 
-export function AccordionProvider(props: {
+interface AccordionProviderProps {
   children: ReactNode;
   unmountChildren?: boolean;
-}) {
-  const { unmountChildren = false } = props;
-  const [state, dispatch] = useReducer(reducer, {
-    items: [],
-    unmountChildren,
-  });
-
-  const utils = useMemo(
-    () => ({
-      change: (title: string, isOpen: boolean) => {
-        return dispatch({
-          type: 'CHANGE_ITEM',
-          payload: { isOpen, title },
-        });
-      },
-      clear: (except: string) => {
-        return dispatch({ type: 'CLEAR', payload: except });
-      },
-      toggle: (title: string) => {
-        return dispatch({ type: 'TOGGLE', payload: title });
-      },
-      remove: (title: string) => {
-        return dispatch({ type: 'REMOVE_ITEM', payload: title });
-      },
+}
+
+export type AccordionActions =
+  | ActionType<'CHANGE', { title: string; isOpen: boolean }>
+  | ActionType<'CLEAR', string>
+  | ActionType<'TOGGLE', string>
+  | ActionType<'REMOVE', string>
+  | ActionType<'CREATE', AccordionItemState>;
+
+function reducer(
+  previous: AccordionItemState[],
+  action: AccordionActions,
+): AccordionItemState[] {
+  return match(action)
+    .with({ type: 'CHANGE' }, ({ payload }) => {
+      const { title, isOpen } = payload;
+      const item = getItem(title, previous);
+
+      if (item) {
+        const itemsWithoutChangedItem = previous.filter(
+          (element) => element.title !== title,
+        );
```

Comment:
I would not put the business logic of identifying an element by its title both in `getItem` and here.

This can be refactored by using a splice or the logic can be moved to the utils.

Also I want to draw your attention to the fact that you changed the order of the elements in this actions and in other actions as well, which was not the case before. This is potentially a regression. Have you checked that we don't rely anywhere on the order?

Reply:
I moved everything to utils functions. I checked everything on the Storybook and he looks to works !
I will re-check if we use something about index or other